### DISCUSSION TITLE
SSR 下的暗色模式 FOUC 的问题

### DISCUSSION THREAD

**Discussion by arvinxx** [ant-design/ant-design#44617] [General]
CSSinJS 通过 SSR/SSG 可以实现单一主题的静态渲染，但针对多主题和响应式场景就会存在问题。譬如使用 nextjs （默认SSR）的应用，一定会 先显示亮色，然后再切换为暗色，会较大程度上影响用户体验。

![efe](https://github.com/ant-design/ant-design/assets/28616219/413e4111-2a95-4de0-be63-daa65c4827e2)

需要有一个方案，在使用 CSSinJS 技术生成样式时来解决这个问题。

目前想到的，可能的思路是在使用 css 变量。在 ssr 的时候分别去生成亮色和暗色下的色值，但具体的实现没啥思路。大家可以讨论下？

**Comment by zombieJ** [comment]
我感觉是做一个 loading 蒙层，好了再打开页面就行了？

**Reply by arvinxx** [reply]
SSR 怎么做？ 不是 CSR，而是 SSR

**Reply by zombieJ** [reply]
感觉聊过这个问题，配置化的主题从 URL 里读取在 SSR 渲染。跟随系统的就没有办法了。

**Reply by arvinxx** [reply]
比如任意一个 SSR 页面，默认出的 HTML 是这样的： 

<img width="1728" alt="image" src="https://github.com/ant-design/ant-design/assets/28616219/05972b21-8142-48d4-9d64-cd8dc47abd7b">

要解决的问题是在首屏 html 出来的时候，就是正常的暗色样式

**Reply by arvinxx** [reply]
> 感觉聊过这个问题，配置化的主题从 URL 里读取在 SSR 渲染。跟随系统的就没有办法了。


如果这个在 SSG 层面是完全无解的，那么CSSinJS 就在比 TailwindCSS ，或者原生的 css 这种要弱一档了… 比如dumi文档站就始终会有这个闪烁的问题

**Reply by Dunqing** [reply]
> 如果这个在 SSG 层面是完全无解的，那么CSSinJS 就在比 TailwindCSS ，或者原生的 css 这种要弱一档了… 比如dumi文档站就始终会有这个闪烁的问题

Automatic CSS 的都是默认包含 light dark 的 css，在 SSR 里把暗色，亮色的css都注入是否能解？

**Reply by arvinxx** [reply]
> Automatic CSS 的都是默认包含 light dark 的 css，在 SSR 里把暗色，亮色的css都注入是否能解？

解不了，因为 html 生成的 className 是指向的 light css 类名

**Comment by MadCcc** [comment]
写一套用 css 变量的主题，在客户端改 css 变量。
由于目前的算法都是需要实时计算的，所以算法部分就需要变成 css 变量映射，另外还要在客户端实现两套 css 变量（亮暗色），可想而知的大工作量。。

**Comment by Dunqing** [comment]
把 token 都转成 css variables，然后你只需要自己管理 dark mode，注入根据 token 生成的 css variables

和这个 unocss 的 preset 处理方式类似 https://github.com/Dunqing/unocss-preset-theme 不知道 antd 的 cssinjs 能不能这么干

**Reply by MadCcc** [reply]
+1，本质上就是把颜色的计算逻辑从算法层下降到原生 js（甚至 build-time），然后直接在客户端注入
